README.md
[![Build Status](https://travis-ci.com/jcs-workspace/swagger-it.svg?branch=master)](https://travis-ci.com/jcs-workspace/swagger-it)
[![Unity Engine](https://img.shields.io/badge/python-%3E=_3.6-green.svg)](https://www.python.org/downloads/)
[![Release Tag](https://img.shields.io/github/tag/jcs-workspace/swagger-it.svg?label=release)](https://github.com/jcs-workspace/swagger-it/releases/latest)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)


# swagger-it
> Parse your microservice project into a [swagger](https://github.com/swagger-api)
yaml file.

This turn your workflow around by code first then generate an update to date
API documentation.

## Installation

**NOTE**: Python 3.6 or higher is required.

```bash
# clone the repo
$ git clone https://github.com/jcs-workspace/swagger-it.git

# change the working directory to sherlock
$ cd sherlock

# install python3 and python3-pip if they are not installed

# install the requirements
python3 -m pip install -r requirements.txt
```

## Supported Languages

* C
* C++
* Objective-C
* C#
* JavaScript
* Java
* Python
* Go
* Ruby

## Todo List

- [ ] Implementes core changes to generate yaml file.
- [ ] Design the template file. (Make it as extensible as possible)
- [ ] Design the tags and docstring to work together.

## References

* https://github.com/swaggo/swag
* https://editor.swagger.io/?_ga=2.130550666.406048183.1579233982-1663214344.1579233982


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.


### 2020-01-20

* Added `guesslang` as a dependencies.

### 2020-01-19

* Added `python-magic-bin` as a dependencies.
* Added `comment_parser` as a dependencies.

### 2020-01-18

* Start the project `swagger-it`.


examples/test.go
/**
 * $File: test.go $
 * $Date: 2020-01-19 20:48:18 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2020 by Shen, Jen-Chieh $
 */



examples/test.js
/**
 * $File: test.js $
 * $Date: 2020-01-19 00:35:46 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2020 by Shen, Jen-Chieh $
 */

"use strict";

/**
 * @swagger
 * @tags accounts
 * @tags.description Everything about your Pets
 * @tags.externalDocs.description Find out more
 * @tags.externalDocs.url http://swagger.io
 */

/**
 * @swagger
 * @summary Show an account
 * @tags accounts
 * @description This is the description to show an account.
 * @accept json
 * @produce json
 * @param id path int true "Account ID"
 * @success 200 object model.Account
 * @failure 400 {object} httputil.HTTPError
 * @failure 404 {object} httputil.HTTPError
 * @failure 500 {object} httputil.HTTPError
 * @verb get
 * @router /accounts/{id}
 */
function showAccount(a, b) {

}

function listAccount(a, b) {

}

/**
 * @swagger
 * @title Swagger Example API
 * @version 1.0
 * @description This is a sample server Petstore server.
 * @termsOfService http://swagger.io/terms/
 *
 * @contact.name API Support
 * @contact.url http://www.swagger.io/support
 * @contact.email support@swagger.io
 *
 * @license.name Apache 2.0
 * @license.url http://www.apache.org/licenses/LICENSE-2.0.html
 *
 * @host petstore.swagger.io
 * @basePath /v2
 *
 * @schemes http, https, ws, wss
 *
 * @externalDocs.description Find out more about Swagger
 * @externalDocs.url http://swagger.io
 */
function main() {

}


info.py
# ========================================================================
# $File: info.py $
# $Date: 2020-01-21 20:53:45 $
# $Revision: $
# $Creator: Jen-Chieh Shen $
# $Notice: See LICENSE.txt for modification and distribution information
#                   Copyright © 2020 by Shen, Jen-Chieh $
# ========================================================================

from util import *

class PropertyInfo(object):
    """Property data structure."""
    _name = ""
    _type = ""
    _description = ""
    _format = ""
    _default = None  # None, True or False

    def __str__(self):
        return f"      {self._name}:\n" + \
               none_string(self._description,
               f"        description: \"{self._type}\"\n") + \
               none_string(self._type,
               f"        type: \"{self._type}\"\n") + \
               none_string(self._format,
               f"        format: \"{self._format}\"\n") + \
               none_string(self._default,
               f"        default: \"{self._default}\"\n")
        pass

class DefinitionInfo(object):
    """Definition data structure."""
    _name = ""  # Variable name.
    _type = ""
    _properties = []

    def __str__(self):
        return f"  {self._name}:\n" + \
               f"    type: {self._type}" + \
               len_zero_string(self._properties, 'properties:\n') + \
               array_to_string(self._properties) + \
               f"    xml:\n" + \
               f"      name: {self._name}"
    pass

class SecurityDefinitionInfo(object):
    """Security definition data structure."""
    _id = ""
    _type = ""
    _name = ""
    _flow = ""
    _authorizationUrl = ""
    _scopes = []  # TODO:  ..
    _in = ""

    def __str__(self):
        return f"  {self._id}:\n" + \
               f"    type: \"{self._type}\"\n" + \
               f"    name: \"{self._name}\"\n" + \
               f"    authorizationUrl: \"{self._authorizationUrl}\"\n" + \
               f"    flow: \"{self._flow}\"\n" + \
               f"    in: \"{self._in}\"\n"
    pass

class ResponseInfo(object):
    """Response data structure."""
    _id = ""  # 400, 404, 405, etc.
    _description = ""
    _type_or_ref = ""

    def type_valid(self):
        """Check _type_or_ref a valid value."""
        return self._type_or_ref == 'array' or \
               self._type_or_ref == 'boolean' or \
               self._type_or_ref == 'integer' or \
               self._type_or_ref == 'number' or \
               self._type_or_ref == 'object' or \
               self._type_or_ref == 'string' or \
               self._type_or_ref == 'file'

    def __str__(self):
        return f"        {self._id}:\n" + \
               f"          description: \"{self._description}\"\n" + \
               none_string(self._type_or_ref,
               "          schema:\n" + \
               none_string(self.type_valid(),
               f"           type: \"{self._type_or_ref}\"\n") + \
               none_string(not self.type_valid(),
               f"           $ref: \"{self._type_or_ref}\"\n"))
    pass

class ParameterInfo(object):
    """Parameter data structure."""
    _name = ""
    _in = ""
    _description = ""
    _required = None
    _ref = ""

    def __str__(self):
        return f"      - name: \"{self._name}\"\n" + \
               f"        in: \"{self._in}\"\n" + \
               none_string(self._description,
               f"        description: \"{self._description}\"\n") + \
               none_string(self._required,
               f"        required: {self._required}\n") + \
               f"        schema:\n" + \
               f"          $ref: \"{self._ref}\"\n"
    pass

class PathInfo(object):
    """Path data structure."""
    _path = ""  # For @router keyword.
    _verb = ""  # post? put?
    _tags = ""
    _summary = ""
    _description = ""
    _operationId = ""
    _consumes = []
    _produces = []
    _parameters = []
    _responses = []
    _security = []

    def __str__(self):
        return f"  {self._path}:\n" + \
               f"    {self._verb}:\n" + \
               f"      tags:\n" + \
               f"      - \"{self._tags}\"\n" + \
               none_string(self._summary,
               f"      summary: \"{self._summary}\"\n") + \
               none_string(self._description,
               f"      description: \"{self._description}\"\n") + \
               none_string(self._operationId,
               f"      operationId: \"{self._operationId}\"\n") + \
               len_zero_string(self._consumes,
               '      consumes:\n') + \
               array_to_string(self._consumes) + \
               len_zero_string(self._produces,
               '      produces:\n') + \
               array_to_string(self._produces) + \
               len_zero_string(self._produces,
               '      parameters:\n') + \
               array_to_string(self._parameters) + \
               len_zero_string(self._responses,
               '      responses:\n') + \
               array_to_string(self._responses) + \
               len_zero_string(self._security,
               '      security:\n') + \
               array_to_string(self._security)

    def add_consume(self, name):
        """Add a new consume."""
        self._consumes.append("- {name}\n")
        pass

    def add_produce(self, name):
        """Add a new produce."""
        self._produces.append("- {name}\n")
        pass

    def add_response(self, id):
        """Add a new response."""
        new_res = ResponseInfo()
        new_res._id = id
        self._responses.append(new_res)
        return new_res

    def get_response(self, id):
        """Return a response data by ID.
        If not found, create a new one."""
        if not id:
            return None
        for response in self._responses:
            if response._id == id:
                return response
            pass
        new_res = self.add_response(id)
        return new_res

    def get_param(self, name):
        """Add a new parameter."""
        # TODO: ..
        pass
    pass

class TagInfo(object):
    """Tag data structure."""
    _name = ""
    _description = ""
    _externalDocs_description = ""
    _externalDocs_url = ""

    def __str__(self):
        return f"- name: \"{self._name}\"\n" + \
               f"  description: \"{self._description}\"\n" + \
               f"  externalDocs:\n" + \
               f"    description: \"{self._externalDocs_description}\"\n" + \
               f"    url: \"{self._externalDocs_url}\"\n"
    pass

class APIInfo(object):
    """Swagger api info section."""
    _description = ""
    _version = ""
    _title = ""
    _termsOfService = ""
    _contact_email = ""
    _license_name = ""
    _license_url = ""

    def __str__(self):
        return f"  description: \"{self._description}\"\n" + \
               f"  version: \"{self._version}\"\n" + \
               f"  title: \"{self._title}\"\n" + \
               f"  termsOfService: \"{self._termsOfService}\"\n" + \
               none_string(self._contact_email,
               f"  contact:\n" + \
               f"    email: \"{self._contact_email}\"\n") + \
               none_string(self._license_name is "" and self._license_url is "",
               f"  license:\n" + \
               f"    name: \"{self._license_name}\"\n" + \
               f"    url: \"{self._license_url}\"\n")
    pass

class SeaggerInfo(object):
    """Swagger API root info."""
    _swagger = "2.0"
    _info = APIInfo()  # Only one instance.
    _host = ""
    _basePath = ""
    _tags = []
    _schemes = []
    _paths = []
    _securityDefinitions = []
    _definitions = []
    _externalDocs_description = ""
    _externalDocs_url = ""

    def __str__(self):
        return f"swagger: \"{self._swagger}\"\n" + \
               f"info:\n{self._info}" + \
               f"host: \"{self._host}\"\n" + \
               f"basePath: \"{self._basePath}\"\n" + \
               len_zero_string(self._tags, 'tags:\n') + \
               array_to_string(self._tags) + \
               len_zero_string(self._schemes, 'schemes:\n') + \
               array_to_string(self._schemes) + \
               len_zero_string(self._paths, 'paths:\n') + \
               array_to_string(self._paths) + \
               "externalDocs:\n" + \
               f"  description: \"{self._externalDocs_description}\"\n" + \
               f"  url: \"{self._externalDocs_url}\""

    def add_scheme(self, option):
        """Scheme defined in Swagger/OpenAPI."""
        if not option is "https" and not option is "http" and \
            not option is "ws" and not option is "wss":
            print('[WARNING] Wrong scheme option:', option)
            return
        self._schemes.append(f"- {option}\n")
        pass

    def add_tag(self, tag_name):
        """Add tag by new TAG_NAME.

        @param { string } tag_name : Name of the tag.
        """
        tag = TagInfo()
        tag._name = tag_name
        self._tags.append(tag)
        return tag

    def get_tag(self, tag_name):
        """Get tag by TAG_NAME.
        If not found, we create a new one."""
        if not tag_name:
            return None
        for tag in self._tags:
            if tag._name == tag_name:
                return tag
            pass
        new_tag = self.add_tag(tag_name)
        return new_tag

    def add_path(self, path_name):
        """Add path by new PATH_NAME.

        @param { string } path_name : Name of the path/route.
        """
        path = PathInfo()
        path._path = path_name
        self._paths.append(path)
        return path

    def get_path(self, path_name):
        """Get path by PATH_NAME.
        If not found, we create a new one."""
        if not path_name:
            return None
        for path in self._paths:
            if path._path == path_name:
                return path
            pass
        new_path = self.add_path(path_name)
        return new_path

    def add_def(self, def_name):
        """"Add one definition by DEF_NAME."""
        defi = DefinitionInfo()
        defi._name = def_name
        self._definitions.append(defi)
        return defi

    def get_def(self, def_name):
        """Get definition by DEF_NAME.

        @param { string } def_name : Name of the definition.
        """
        if not def_name:
            return None
        for defi in self._definitions:
            if defi._name == def_name:
                return defi
            pass
        new_defi = self.add_def(def_name)
        return new_defi
    pass


util.py
# ========================================================================
# $File: util.py $
# $Date: 2020-01-20 15:44:44 $
# $Revision: $
# $Creator: Jen-Chieh Shen $
# $Notice: See LICENSE.txt for modification and distribution information
#                   Copyright © 2020 by Shen, Jen-Chieh $
# ========================================================================

import pathlib

def warn_exists(id, checker):
    """Warn the non-existence for ID without the CHECKER exists."""
    print(f'[WARNING] Defined `{id}` without `{checker}` defined')
    pass

def safe_get_value(lst, index):
    """Safe get the value from a LST with INDEX."""
    if index >= 0 and index < len(lst):
        return lst[index]
    return None

def none_string(vr, def_str, none_str = ""):
    """Return DEF_STR or NONE_STR depends on weather VR is None or empty string.

    @param { any } vr : Any variable you want to check.
    @param { string } def_str : Default string.
    @param { string } none_str : None string.
    """
    if vr is None or vr is "" or vr is False:
        return none_str
    return def_str

def len_zero_string(lst, def_str, zero_str = ""):
    """Return DEF_STR or ZERO_STR depends on LST's length is 0 or not.

    @param { Array } lst : List to get the length.
    @param { string } def_str : Default string.
    @param { string } zero_str : Zero string.
    """
    if len(lst) is 0:
        return zero_str
    return def_str

def dict_get_value(dic, key):
    """Get the key value.

    @param { Dictionary } dic : Dictionary use to get.
    @param { string } key : Key to get value.
    """
    if dict_key_exists(dic, key):
        return dic[key]
    return None

def dict_key_exists(dic, key):
    """Check if the KEY exist in DIC.

    @param { Dictionary } dic : Dictionary use to check.
    @param { string } key : Key to check.
    """
    return key in dic

def get_file_extension(path):
    """Return the file PATH extension.

    @param { string } path : Target file path we are going to extract.
    """
    return pathlib.Path(path).suffix

def array_to_string(lst):
    """Convert LST to string.

    @param { Array } lst : List of object.
    """
    form_str = ""
    for item in lst:
        form_str += str(item)
        pass
    return form_str

def contain_in_list_equal(item, lst):
    """Check if ITEM is one of the item in LST.

    @param { string } item : Item to check.
    @param { Array } lst : List of item to check.
    """
    for cur_item in lst:
        if cur_item == item:
            return True
    return False

def contain_in_list(item, lst):
    """Check if each item in LST inside the ITEM.

    @param { string } item : Path for major check.
    @param { Array } lst : List of string to check.
    """
    for cur_item in lst:
        if cur_item in item:
            return True
    return False

class ArgumentError(LookupError):
    """Argument input error."""
    pass

def mkdir_safe(path):
    """Make directory if not exists.

    @param { string } path : Directory path.
    """
    pathlib.Path(path).mkdir(parents=True, exist_ok=True)
    pass

def args_input(input):
    """Return the default input path.

    @param { string } input : Default input argument.
    """
    if input is None:
        return os.path.dirname(os.path.abspath(__file__))
    return input


src/main.py
# ========================================================================
# $File: main.py $
# $Date: 2020-01-17 16:04:44 $
# $Revision: $
# $Creator: Jen-Chieh Shen $
# $Notice: See LICENSE.txt for modification and distribution information
#                   Copyright © 2020 by Shen, Jen-Chieh $
# ========================================================================

from guesslang import Guess

def main():
    """Program Entry point.
    """

    name = Guess().language_name("""
    % Quick sort

      -module (recursion).
      -export ([qsort/1]).

      qsort([]) -> [];
      qsort([Pivot|T]) ->
             qsort([X || X <- T, X < Pivot])
             ++ [Pivot] ++
             qsort([X || X <- T, X >= Pivot]).
    """)


    print(name)  # >>> Erlang

if __name__ == "__main__":
    main()


swagger-it.py
# ========================================================================
# $File: swagger-it.py $
# $Date: 2020-01-17 16:04:44 $
# $Revision: $
# $Creator: Jen-Chieh Shen $
# $Notice: See LICENSE.txt for modification and distribution information
#                   Copyright © 2020 by Shen, Jen-Chieh $
# ========================================================================

import os
import platform
import pathlib
import re

from argparse import ArgumentParser, RawDescriptionHelpFormatter
from comment_parser import comment_parser
from util import *
from info import *


module_name = "swagger-it: Parse your microservice project into a swagger yaml file."
__version__ = "0.0.1"

swagger_identifier = "@swagger"

swagger_ids = [# -- Info ---------------------------------------------------
               "@description", "@version", "@title", "@termsOfService",
               "@contact.email", "@license.name", "@license.url",
               "@host", "@basePath", "@schemes",
               "@externalDocs.description", "@externalDocs.url",
               # -- Tags ---------------------------------------------------
               "@tags",  # Followed by name.
               "@tags.description",
               "@tags.externalDocs.description", "@tags.externalDocs.url",
               # -- Paths --------------------------------------------------
               "@router",  # Followed by name.
               "@verb",  # For GET, POST, PUT, DELETE, etc.
               "@summary", "@description",
               "@operationId",
               "@param",
               "@response", "@success", "@failure",
               # -- Security Definitions -----------------------------------
               "@sec_def",  # Followed by name.
               "@name",
               "@type", "@authorizationUrl", "@flow", "@in",
               "@scope",
               # -- Definitions --------------------------------------------
               "@def",  # Followed by name.
               "@type",
               "@property.description", "@property.type", "@property.format"]

ignore_dir = ['.git', '.vs', '.vscode', '.log', 'node_modules', '__pycache__'];

# Valid programminga language; hence we won't waste our time parsing and
# analyzing the source file we don't want.
valid_ext = ['.c', '.h',
             '.cpp', '.hpp', '.hin', '.cin',
             '.m',
             '.cs', '.js', '.java',
             '.py', '.go', '.rb']
mime_type = ['text/x-c', 'text/x-c',
             'text/x-c++', 'text/x-c++', 'text/x-c++', 'text/x-c++',
             'text/x-c++',
             'text/x-c++', 'application/javascript', 'text/x-java-source',
             'text/x-python', 'text/x-go', 'text/x-ruby']

# The swagger info buffer/file.
# This is the final output result, everything and data will store here.
swagger_info = SeaggerInfo()

exists_tag = None
exists_path = None


def get_mime_type_by_extension(ext):
    """Return the mime type string by extension (EXT).

    @param { string } ext : Extension string.
    """
    index = 0
    id = -1
    for vd_ext in valid_ext:
        if vd_ext == ext:
            id = index
            break
        index += 1
        pass
    if id == -1:
        raise ArgumentError('[ERROR] Invalid extension for mime type:', ext)
    return mime_type[id]

def valid_source_file(path):
    """Check if the file PATH valid source file.

    @param { string } path : File path to check valid.
    """
    ext = get_file_extension(path)
    return contain_in_list_equal(ext, valid_ext)

def extract_swagger_identifier(lst_comment):
    """Extract swagger docstring from list of comment/docstring.

    @param { Array } lst_comment : List of comment/docstring.
    """
    valid_swagger_comments = []
    for comment in lst_comment:
        if swagger_identifier in comment.text():
            valid_swagger_comments.append(comment)
            pass
        pass
    return valid_swagger_comments

def form_attribute_list(lst_comment):
    """Turn a list of comments/docstrings to key value pair like JSON.

    @param { typename } lst_comment : List of comment/docstring that contain
    swagger identifier.

    @example
      - @swagger : 2.0
      - @schemes : http, https, ws, wss
      ...
    """
    attr_pair_lst = []
    for comment in lst_comment:
        comment = comment.text().replace('\n', ' ')
        comment = comment.replace(' * ', '')
        attr_lst = comment.split(" @")

        attr = {}

        for attr_line in attr_lst:
            # Form the attribute ID.
            attr_pair = attr_line.split(' ')
            attr_id = '@' + attr_pair.pop(0)
            attr_value = ' '.join(attr_pair)

            if not contain_in_list_equal(attr_id, swagger_ids):
                continue

            attr[attr_id] = attr_value.strip()
            pass
        if attr:
            attr_pair_lst.append(attr)
            pass
        pass
    return attr_pair_lst

def fill_info(key, value):
    """Fill in the swagger info by KEY and VALUE.

    @param { string } key : Attribute key.
    @param { string } value : Attribute value.
    """
    # -- Info ---------------------------------------------------
    if key == '@description':
        if exists_path:  # @router exists.
            swagger_info.get_path(exists_path)._description = value
            pass
        else:
            swagger_info._info._description = value
    elif key == '@version':
        swagger_info._info._version = value
    elif key == '@title':
        swagger_info._info._title = value
    elif key == '@termsOfService':
        swagger_info._info._termsOfService = value
    elif key == '@contact.email':
        swagger_info._info._contact_email = value
    elif key == '@license.name':
        swagger_info._info._license_name = value
    elif key == '@license.url':
        swagger_info._info._license_url = value
    elif key == '@host':
        swagger_info._host = value
    elif key == '@basePath':
        swagger_info._basePath = value
    elif key == '@schemes':
        there_http = re.search('http[^s]', value)
        there_https = re.search('https', value)
        there_ws = re.search('ws[^s]', value)
        there_wss = re.search('wss', value)
        if there_http: swagger_info.add_scheme('http')
        if there_https: swagger_info.add_scheme('https')
        if there_ws: swagger_info.add_scheme('ws')
        if there_wss: swagger_info.add_scheme('wss')
    elif key == 'externalDocs.description':
        swagger_info._externalDocs_description = value
    elif key == 'externalDocs.url':
        swagger_info._externalDocs_url = value
    elif key == '@externalDocs.description':
        swagger_info._externalDocs_description = value
    elif key == '@externalDocs.url':
        swagger_info._externalDocs_url = value
    # -- Tags ---------------------------------------------------
    elif key == '@tags':
        swagger_info.get_tag(value)  # Create one on it's own
        pass
    elif key == '@tags.description':
        if not exists_tag:
            warn_exists('@tags.description', '@tags')
        else:
            swagger_info.get_tag(exists_tag)._description = value
    elif key == '@tags.externalDocs.description':
        if not exists_tag:
            warn_exists('@tags.externalDocs.description', '@tags')
        else:
            swagger_info.get_tag(exists_tag)._externalDocs_description = value
    elif key == '@tags.externalDocs.url':
        if not exists_tag:
            warn_exists('@tags.externalDocs.url', '@tags')
        else:
            swagger_info.get_tag(exists_tag)._externalDocs_url = value
    # -- Paths ---------------------------------------------------
    elif key == '@router':
        swagger_info.get_path(value)  # Create one on it's own

        # TODO: Parse {id} etc. And make one parameter.
    elif key == '@verb':
        if not exists_path:
            warn_exists('@verb', '@router')
        else:
            swagger_info.get_path(exists_path)._verb = value
    elif key == '@summary':
        if not exists_path:
            warn_exists('@summary', '@router')
        else:
            swagger_info.get_path(exists_path)._summary = value
    elif key == '@operationId':
        if not exists_path:
            warn_exists('@operationId', '@router')
        else:
            swagger_info.get_path(exists_path)._operationId = value
    elif key == '@param':
        pass
    elif key == '@response' or key == '@success' or key == '@failure':
        if not exists_path:
            warn_exists('@response / @success / @failure', '@router')
        else:
            fmt_val_lst = value.split(' ')
            current_path = swagger_info.get_path(exists_path)

            # This is where I parse `@response` format.
            res_id = safe_get_value(fmt_val_lst, 0)
            res_type_or_ref = safe_get_value(fmt_val_lst, 1)
            res_desc = safe_get_value(fmt_val_lst, 2)

            new_res = current_path.get_response(res_id)
            new_res._description = res_desc
            new_res._type_or_ref = res_type_or_ref
        pass
    # -- Security Definitions -----------------------------------
    elif key == '@securityDefinitions':
        pass
    # -- Definitions --------------------------------------------
    elif key == '@definitions':
        pass
    else:
        print('[WARNING] Use key not found:', key)
        pass
    pass

def fill_swagger_info(attr_lst):
    """Fill the swagger information by attribute list (ATTR_LST).

    @param { Array } attr_lst : List of attributes.
    """
    global exists_tag
    global exists_path
    for attr in attr_lst:
        exists_tag = dict_get_value(attr, '@tags')
        exists_path = dict_get_value(attr, '@router')
        for key in attr:
            fill_info(key, attr[key])
            pass
        pass
    pass

def form_swagger_buffer(comments):
    """Form the swagger buffer.

    @param { Array } comments : List of comments or properly swagger docstrings.
    """
    for comment in comments:

        pass
    return template_buffer

def main():
    """Program Entry point."""

    version_string = f"%(prog)s {__version__}\n"

    parser = ArgumentParser(formatter_class=RawDescriptionHelpFormatter,
                            description=f"{module_name} (Version {__version__})")

    parser.add_argument("--input", "-i", dest="input",
                        help="Input path, this can be a directory or file.")

    parser.add_argument("--output", "-o", dest="output",
                        help="Output path, this must be a path point to a file.")

    parser.add_argument("--version",
                        action="version",  version=version_string,
                        help="Display version information and dependencies.")

    args = parser.parse_args()

    input = args_input(args.input)
    output = args.output

    if output is None:
        dirname = output   # Set it to None.
        filename = output  # Set it to None.
    else:
        dirname = os.path.dirname(output)    # Target directory path.
        filename = os.path.basename(output)  # Target file to export as swagger yaml.
        pass

    isFile = os.path.isfile(input)
    isDir = os.path.isdir(input)

    print("[SWAGGER-IT] Start checking input:::")

    if not isFile and not isDir:
        raise ArgumentError('[ERROR] File input error:', input)

    print("[SWAGGER-IT] Start checking output:::")

    if not dirname and not filename:
        raise ArgumentError('[ERROR] File output error:', output)

    print("[SWAGGER-IT] Parsing comments and docstrings:::")

    comments = []

    if isFile:  # When is file..
        extension = get_file_extension(input)
        mime_text = get_mime_type_by_extension(extension)
        comments = comment_parser.extract_comments(input, mime=mime_text)
    else:  # When is directory..
        for r, d, f in os.walk(input):
            for file in f:
                filepath = os.path.join(r, file)
                filepath = filepath.replace("\\", "/")
                if not contain_in_list(filepath, ignore_dir) and \
                   valid_source_file(filepath):
                    extension = get_file_extension(filepath)
                    mime_text = get_mime_type_by_extension(extension)
                    new_comments = comment_parser.extract_comments(filepath, mime=mime_text)
                    comments.extend(new_comments)
                    pass
                pass
            pass
        pass

    comments = extract_swagger_identifier(comments)
    attr_lst = form_attribute_list(comments)

    print(comments)

    print("[SWAGGER-IT] Filling swagger informations:::")

    fill_swagger_info(attr_lst)

    print(swagger_info)

    mkdir_safe(dirname)  # Ensure the path exists.

    file = open(output, "w+")
    file.write(str(swagger_info))
    file.close()

    print("[SWAGGER-IT] Done generate the file:::")
    pass

if __name__ == "__main__":
    main()
    pass
